### PR TITLE
Add portable_simd support

### LINKED ISSUE #33: Migrate from packed_simd to stdlib simd (aka portable simd)
Packed simd stopped compiling a few days ago: https://github.com/rust-lang/packed_simd/issues/343. It should be replaced with portable simd: https://doc.rust-lang.org/std/simd/index.html

### PR DESCRIPTION
**Pull request by farnoy**
Fixes #33

Does not have acceleration for trigonometric functions due to https://github.com/rust-lang/portable-simd/issues/6

Follows the `packed_simd` implementation closely, but masks are called `mask<$elt>x<$lanes>` not `m<$elt>x<$lanes>`

Has a few more things than packed_simd like native `recip`, `floor`, `ceil`, `trunc`, but crucially it's missing trigonometry accelration.

### CODE REVIEW

File: src/simd/portable_simd_impl.rs [hunk 1]
```diff
@@ -0,0 +1,1686 @@
...
+            }
+
+            #[inline(always)]
+            fn extract(&self, i: usize) -> Self::Element {
+                self.0[i]
+            }
+
+            #[inline(always)]
+            unsafe fn extract_unchecked(&self, i: usize) -> Self::Element {
+                self.0[i]
```
Comment (farnoy) [inline src/simd/portable_simd_impl.rs:320]: There's no unchecked version from what I've seen

File: src/simd/portable_simd_impl.rs [hunk 2]
```diff
@@ -0,0 +1,1686 @@
...
+            }
+
+            #[inline(always)]
+            fn replace(&mut self, i: usize, val: Self::Element) {
+                self.0[i] = val;
+            }
+
+            #[inline(always)]
+            unsafe fn replace_unchecked(&mut self, i: usize, val: Self::Element) {
+                self.0[i] = val;
```
Comment (farnoy) [inline src/simd/portable_simd_impl.rs:330]: Same here?

File: src/simd/portable_simd_impl.rs [hunk 3]
```diff
@@ -0,0 +1,1686 @@
...
+//! Traits for SIMD values.
+
+use crate::scalar::{ComplexField, Field, SubsetOf, SupersetOf};
+use crate::simd::{
+    PrimitiveSimdValue, SimdBool, SimdComplexField, SimdPartialOrd, SimdRealField, SimdSigned,
+    SimdValue,
+};
+use approx::AbsDiffEq;
+#[cfg(feature = "decimal")]
+use decimal::d128;
```
Comment (farnoy) [inline src/simd/portable_simd_impl.rs:13]: Have not tested with this feature, TODO

File: src/simd/rand_impl.rs [hunk 4]
```diff
@@ -200,3 +200,91 @@ impl_rand_packed_simd!(
...
 );
+
+#[cfg(feature = "portable_simd")]
+macro_rules! impl_rand_portable_simd(
+    ($($wrapped: ty, $($i: ident),*;)*) => ($(
+        impl rand::distributions::Distribution<$wrapped> for rand::distributions::Standard {
+            #[inline(always)]
+            fn sample<R: rand::Rng + ?Sized>(&self, rng: &mut R) -> $wrapped {
+                <$wrapped>::new($(
+                    ignore_snd!([self.sample(rng)], [$i])
```
Comment (farnoy) [inline src/simd/rand_impl.rs:211]: There's a nicer specialization for `packed_simd` vectors in the `rand` crate, not sure if it's faster or by how much. It uses fill_bytes loaded into an int vector and then shuffled to fit into a float vector.

**Comment by sebcrozet** [comment]
Thank you for this PR! Since it’s been a while, I took the liberty of rebasing on top of `master` and updated to the latest nightly.

**Pull request merged**